messaging/src/main/java/org/axonframework/serialization/avro/AvroSerializer.java

```diff
@@ -0,0 +1,247 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.serialization.avro;
+
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.message.SchemaStore;
+import org.apache.avro.util.ClassUtils;
+import org.axonframework.common.AxonConfigurationException;
+import org.axonframework.serialization.ChainingConverter;
+import org.axonframework.serialization.Converter;
+import org.axonframework.serialization.RevisionResolver;
+import org.axonframework.serialization.SerializedObject;
+import org.axonframework.serialization.SerializedType;
+import org.axonframework.serialization.Serializer;
+import org.axonframework.serialization.SimpleSerializedType;
+import org.axonframework.serialization.UnknownSerializedType;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Serializer providing support for Apache Avro and using Single Object Encoded binary encoding.
+ */
+public class AvroSerializer implements Serializer {
+
+    private final RevisionResolver revisionResolver;
+    private final ChainingConverter converter = new ChainingConverter();
+    private final List<AvroSerializerStrategy> serializerStrategies = new ArrayList<>();
+    /*
+     * Responsible for everything that is NOT Avro (e.g. MetaData).
+     */
+    private final Serializer serializerDelegate;
+
+    /**
+     * Creates the serializer instance.
+     * @param builder builder containing relevant settings.
+     */
+    protected AvroSerializer(@Nonnull Builder builder) {
+        builder.validate();
+        this.revisionResolver = builder.revisionResolver;
+        this.serializerDelegate = builder.serializerDelegate;
+        this.serializerStrategies.addAll(builder.serializerStrategies);
+        this.serializerStrategies.add(new SpecificRecordBaseSerializerStrategy(
+                                              builder.schemaStore,
+                                              this.revisionResolver
+                                      )
+        );
+        this.converter.registerConverter(new ByteArrayToGenericRecordConverter(builder.schemaStore));
+    }
+
+    /**
+     * Creates a builder for Avro Serializer.
+     * @return fluent builder instance.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> SerializedObject<T> serialize(Object object, @Nonnull Class<T> expectedRepresentation) {
+        // assume: expectedRepresentation is byte[] (or possibly String), nothing else.
+        // TODO: Question is this correct? What is the contract of null handling?
+        Objects.requireNonNull(object, "Can't serialize a null object");
```

Comment:
Valid. Since there is a inconsistency in other serializers.

What is your expectation?

We can of course support `null` by returning an emptySerializedObject...

Reply:
I mean there is an asymmetry between serialize/deserialize

Reply:
Json serializer test creates a null object and serializes it. The Jackson JSON representation of this null is "null". I believe this is not valid JSON, even if Jackson does it this way.

Reply:
We will delegate to serializerDelegate :-)